File: data_structure/1_linked_list/5_remove_nth/remove_nth.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        def findNodeBeforeNth (head: ListNode, n: Int) -> ListNode:
            nodeIndexBeforeNth = n+1
            current = head
            nodeBeforeNth = head
            for i in range(0, nodeIndexBeforeNth):
                if not current:
                    return None
                current = current.next

            while current:
                current = current.next
                nodeBeforeNth = nodeBeforeNth.next

            return nodeBeforeNth

        def deleteNode (node1, node2):
            node1.next = node2.next

        nodeBeforeNth = findNodeBeforeNth(head, n)

        if not nodeBeforeNth:
            return head.next
        else:
            deleteNode(nodeBeforeNth, nodeBeforeNth.next)
            return head;

File: data_structure/1_linked_list/5_remove_nth/remove_nth.js
'use strict';

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {

  function findNodeBeforeNth(head, n){
    let nodeIndexBeforeNth = n+1;
    let current = head;
    let nodeBeforeNth = head;
    for (let i=0; i<nodeIndexBeforeNth; i++){
      if (!current) return null;
      current = current.next;
    }
    while (current){
      current = current.next;
      nodeBeforeNth = nodeBeforeNth.next;
    }
    return nodeBeforeNth;
  }

  function deleteNode (node1, node2){
    node1.next = node2.next;
  }

  let nodeBeforeNth = findNodeBeforeNth(head, n);
  if (!nodeBeforeNth) return head.next;
  else {
    deleteNode(nodeBeforeNth,nodeBeforeNth.next);
    return head;
  }
};


module.exports = removeNthFromEnd;
